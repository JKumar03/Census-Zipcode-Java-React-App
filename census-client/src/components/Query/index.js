import React from "react";
import { useDispatch } from "react-redux";
import { Accordion, Button, Table, Alert, FormControl, Card } from "react-bootstrap";
import { QUERIES } from "../../actions/queries";

const handleClick = (query, dispatch) => {
  // Dispatch the action to run the API query
  dispatch({ type: QUERIES.RUN_QUERY, payload: query });
}

const getFileParam = (e) => {
  // Append form data for file object
  const formData = new FormData();
  formData.append('file', e.target.files[0], e.target.files[0].name);
  return formData;
}

const handleChange = (e, queryId, dispatch) => {
  // Dispatch the action to update the parameter in corresponding API query
  dispatch({
    type: QUERIES.UPDATE_PARAM,
    payload: {
      queryId,
      name: e.target.name,
      value: e.target.name === 'file' ? getFileParam(e) : e.target.value
    }
  });
}

const Query = ({ query }) => {
  const { title, endpoint, params = [] } = query;
  const dispatch = useDispatch();

  return (
    <Card>
      <Accordion.Toggle as={Card.Header} eventKey={query.queryId}>
        <strong>{title}</strong>
      </Accordion.Toggle>
      <Accordion.Collapse eventKey={query.queryId}>
        <Card.Body>
          <strong>URL Endpoint:</strong><br />
          <strong className="method text-success">{query.method}</strong> <span className="endpoint">{endpoint}</span>
          <br /><br />
          <strong>Parameters</strong>
          {!Object.keys(params).length
            // Show placeholder if there are no parameters
            ? (
              <Alert variant="info">
                No Parameters
              </Alert>
            )
            // Otherwise, show parameters table
            : (
              <Table bordered>
                <thead>
                  <tr>
                    <td>Parameter</td>
                    <td>Value</td>
                  </tr>
                </thead>
                <tbody>
                  {
                    // Loop over the params object and show input fields
                    Object.keys(params).map((p, index) => (
                      <tr key={index}>
                        <td>{p}</td>
                        <td>
                          {p === 'file'
                            ? <FormControl
                              type="file"
                              onChange={(e) => handleChange(e, query.queryId, dispatch)}
                              name={p}
                            />
                            : <FormControl
                              onChange={(e) => handleChange(e, query.queryId, dispatch)}
                              value={params[p]}
                              name={p}
                            />
                          }
                        </td>
                      </tr>
                    ))}
                </tbody>
              </Table>
            )
          }
          <Button variant="primary" onClick={() => handleClick(query, dispatch)}>Try it!</Button>
        </Card.Body>
      </Accordion.Collapse>
    </Card>
  );
}

export default Query;