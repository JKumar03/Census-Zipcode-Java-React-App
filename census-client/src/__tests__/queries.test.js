import React from 'react';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "../store";

import Query from '../components/Query';

describe('Query component:', () => {
  let query = null;
  let component = null;
  const store = configureStore();

  beforeEach(() => {
    query = {
      queryId: 1,
      title: 'Sample title',
      endpoint: "http://example.com",
      method: 'GET',
      params: {
        dummy: 1
      }
    };
    component = mount(
      <Provider store={store}>
        <Query query={query} />
      </Provider>
    );
  });

  it('should match the snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('displays the correct title', () => {
    expect(component.find('.card-header').text()).toEqual(query.title);
  });

  it('displays the correct endpoint', () => {
    expect(component.find('.card-body .endpoint').text()).toEqual(query.endpoint);
  });

  it('displays the correct params', () => {
    expect(component.find('.card-body input')).toHaveLength(Object.keys(query.params).length);
  });

  it('displays the correct method', () => {
    expect(component.find('.card-body .method').text()).toEqual(query.method);
  });
});
