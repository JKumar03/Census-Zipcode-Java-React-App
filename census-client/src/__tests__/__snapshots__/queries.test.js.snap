// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query component: should match the snapshot 1`] = `ReactWrapper {}`;
